### PR TITLE
TLS Session ticket support

### PR DESCRIPTION
**Pull request by redbrain17**
- fixed issues of review
- rebased with current master

don't take the two former pull request. I think I messed up these. Please review only this one

### CODE REVIEW

**Comment by jay** [comment]
Closed #3058 #3059 in favor of this one. In the future if you need to change the commits please force-push to your branch rather than start a new pull request with a new branch. Session tickets have been proposed before but the idea usually petered off, I'm not sure why. Is there some advantage to using them over session id which is already supported?

File: lib/vtls/nss.c [hunk 1]
```diff
@@ -2029,6 +2029,11 @@ static CURLcode nss_setup_connect(struct connectdata *conn, int sockindex)
   }
 #endif
 
+   if (data->set.ssl.sessionticket) {
+           if (SSL_OptionSet(BACKEND->handle, SSL_ENABLE_SESSION_TICKETS, PR_TRUE) != SECSuccess) {
+               goto error;
+           }
```
Comment (bagder) [inline lib/vtls/nss.c]: wrong indent level

File: src/tool_getparam.c [hunk 2]
```diff
@@ -1659,6 +1660,10 @@ ParameterError getparameter(const char *flag, /* f or -long-flag */
           config->socks5_auth &= ~CURLAUTH_GSSAPI;
         break;
 
+     case 'C': /* --tls-use-session-tickets */
+           config->tls_use_session_ticket = TRUE;
+		break;
```
Comment (bagder) [inline src/tool_getparam.c]: wrong indent level

File: src/tool_help.c [hunk 3]
```diff
@@ -332,6 +332,8 @@ static const struct helptxt helptext[] = {
    "TLS 1.3 proxy cipher suites"},
   {"    --proxy-tlsauthtype <type>",
    "TLS authentication type for HTTPS proxy"},
+  {"    --tls-use-session-tickets",
+   "Enable TLS session tickets (NSS)"},
```
Comment (bagder) [inline src/tool_help.c]: note that this list is generated (see comment above the list), so the contents need to be in files in `docs/cmdline-opts/` - you need to create a new file there for a new option

File: src/tool_operate.c [hunk 4]
```diff
@@ -1482,6 +1482,10 @@ static CURLcode operate_do(struct GlobalConfig *global,
           my_setopt(curl, CURLOPT_PROXY_SSL_OPTIONS,
                     (long)CURLSSLOPT_ALLOW_BEAST);
 
+	if (config->tls_use_session_ticket) {
+          my_setopt(curl, CURLOPT_TLS_USE_SESSION_TICKETS, 1L);
+        }
```
Comment (bagder) [inline src/tool_operate.c]: wrong indent level

**Comment by bagder** [comment]
Could we consider simple using a ticket if `CURLOPT_SSL_SESSIONID_CACHE` is true instead of adding a new option? Will there really be a situation where a user wants one but not the other?

**Comment by bagder** [comment]
... or perhaps convert it to a bitmask and add a bitflag for ticket?

**Comment by bagder** [comment]
Also, pay extra attention to the [test 1119 and 1139 failures](https://travis-ci.org/curl/curl/jobs/434696188#L5827).

**Comment by redbrain17** [comment]
the difference between TLS session ids and TLS Session Tickets are explained here:

https://stackoverflow.com/questions/19939247/ssl-session-tickets-vs-session-ids
according to this link there exits indeed situations where tickets are desirable over session identifiers. (RFC 5077)
So I would suggest an extra option here.

**Comment by jay** [comment]
Has anyone done a real-world comparison of the proportion of servers that support each method?

**Comment by bagder** [comment]
> So I would suggest an extra option here.

What about instead making bit 1 of the argument to`CURLOPT_SSL_SESSIONID_CACHE` become a flag to toggle session tickets? The option only has a single bit defined currently so it would not break anything but will save us from adding another option. The session_id is then not entirely correctly named but is still a function that is very similar in function.

**Comment by mkauf** [comment]
#2645 is another old version of this PR.

**Comment by danielgustafsson** [comment]
Ping @redbrain17. Have you had time to work on addressing the comments made on this PR?

**Comment by redbrain17** [comment]
On 24/12/18, Daniel Gustafsson wrote:
> Ping @redbrain17. Have you had time to work on addressing the comments made on this PR?

still alive and interested to get this merged. I'll look in the next
days on it to improve it.  

Best Regards,

Oliver

**Comment by danielgustafsson** [comment]
> On 26 Dec 2018, at 11:45, Oliver Graute <notifications@github.com> wrote:
> 
> On 24/12/18, Daniel Gustafsson wrote:
> > Ping @redbrain17. Have you had time to work on addressing the comments made on this PR?
> 
> still alive and interested to get this merged. I'll look in the next
> days on it to improve it. 

Great! Let us know if you need any help.

File: src/tool_help.c (hunk 3 above)
Comment (redbrain17) [reply src/tool_help.c]: ok I added a file tls-sessiontickets.d should I drop this two lines then?

File: src/tool_operate.c (hunk 4 above)
Comment (redbrain17) [reply src/tool_operate.c]: do  `make checksrc` find such indent level errors?

**Comment by redbrain17** [comment]
improved the patches. Ready for next review.

File: src/tool_getparam.c [hunk 5]
```diff
@@ -1670,6 +1671,11 @@ ParameterError getparameter(const char *flag, /* f or -long-flag */
           config->socks5_auth &= ~CURLAUTH_GSSAPI;
         break;
 
+      case 'C':
+        /* --tls-use-session-tickets */
+        config->tls_use_session_ticket = TRUE;
```
Comment (bagder) [inline src/tool_getparam.c]: this should be `= toggle` so that the `--no-` prefix works.

File: src/tool_getparam.c [hunk 6]
```diff
@@ -264,6 +264,7 @@ static const struct LongShort aliases[]= {
   {"E9", "proxy-tlsv1",              ARG_NONE},
   {"EA", "socks5-basic",             ARG_BOOL},
   {"EB", "socks5-gssapi",            ARG_BOOL},
+  {"EC", "tls-use-session-tickets",  FALSE},
```
Comment (bagder) [inline src/tool_getparam.c]: Use `ARG_BOOL` instead of `FALSE`.

File: lib/vtls/nss.c [hunk 7]
```diff
@@ -2070,6 +2070,11 @@ static CURLcode nss_setup_connect(struct connectdata *conn, int sockindex)
   }
 #endif
 
+   if (data->set.ssl.sessionticket) {
+           if (SSL_OptionSet(BACKEND->handle, SSL_ENABLE_SESSION_TICKETS, PR_TRUE) != SECSuccess) {
```
Comment (bagder) [inline lib/vtls/nss.c]: Is there a particular reason to have this feature default to off?

File: src/tool_help.c (hunk 3 above)
Comment (bagder) [reply src/tool_help.c]: No, they still need to be there. but my comment about "(NSS)" applies here too...

File: docs/cmdline-opts/tls-sessiontickets.d [hunk 8]
```diff
@@ -0,0 +1,7 @@
+Long: tls-use-session-tickets
+Arg:
+Help: Enable TLS session tickets (NSS)
+Added: 7.6x.0
+---
+Set TLS Session Tickets in NSS for Session Resumption without Server-Side
+State (RFC 5077)
```
Comment (bagder) [inline docs/cmdline-opts/tls-sessiontickets.d]: I think this explanation should expand a little on why a user would like to use this option even if they're not experts in TLS protocol details.

Perhaps drop "use" part from the flag name to shorten it somewhat?

File: include/curl/curl.h [hunk 9]
```diff
@@ -1894,6 +1894,9 @@ typedef enum {
   /* set this to 1L to allow HTTP/0.9 responses or 0L to disallow */
   CINIT(HTTP09_ALLOWED, LONG, 285),
 
+  /* TLS session tickets*/
+  CINIT(TLS_USE_SESSION_TICKETS, LONG, 286),
```
Comment (bagder) [inline include/curl/curl.h]: I think perhaps this should rather be a new bit in the `CURLOPT_SSL_OPTIONS` option. What do you think?

It also needs to be documented. And added to `docs/libcurl/symbols-in-versions`.

File: src/tool_help.c (hunk 3 above)
Comment (redbrain17) [reply src/tool_help.c]: Missed you comment according (NSS) So I dropped it. But now its not clear that RFC 5077 is only supported by NSS implementation of TLS or I'am wrong here?

File: src/tool_getparam.c (hunk 5 above)
Comment (redbrain17) [reply src/tool_getparam.c]: done

File: docs/cmdline-opts/tls-sessiontickets.d (hunk 8 above)
Comment (redbrain17) [reply docs/cmdline-opts/tls-sessiontickets.d]: done

**Comment by bagder** [comment]
Abandoned. Closing.

**Pull request closed without merging**